pandaserver/daemons/scripts/add_main.py

```diff
@@ -0,0 +1,326 @@
+                        adder_gen = AdderGen(taskBuffer, panda_id, job_status, attempt_nr,
+                                       ignoreTmpError=False, siteMapper=aSiteMapper, pid=uniq_pid, prelock_pid=prelock_pid)
+                    elif (timeInt - modTime) > datetime.timedelta(minutes=gracePeriod):
+                        # add
+                        # tmpLog.debug("Add File {0} : {1}".format(os.getpid(),fileName))
+                        # thr = AdderGen(taskBuffer,match.group(1),match.group(2),fileName,
+                        #                ignoreTmpError=True,siteMapper=aSiteMapper)
+                        tmpLog.debug("Add pid={0} job={1}.{2} st={3}".format(uniq_pid, panda_id, attempt_nr, job_status))
+                        adder_gen = AdderGen(taskBuffer, panda_id, job_status, attempt_nr,
+                                       ignoreTmpError=True, siteMapper=aSiteMapper, pid=uniq_pid, prelock_pid=prelock_pid)
+                    if adder_gen is not None:
+                        adder_gen.run()
+                        del adder_gen
+                except Exception:
+                    type, value, traceBack = sys.exc_info()
+                    tmpLog.error("%s %s" % (type,value))
+            # unlock prelocked reports if possible
+            for panda_id, job_status, attempt_nr, time_stamp in prelocked_JOR_list:
+                taskBuffer.unlockJobOutputReport(
+                            panda_id=panda_id, attempt_nr=attempt_nr, pid=prelock_pid)
+            # close taskBuffer connection
+            while True:
+                try:
+                    proxy = taskBuffer.proxyPool.proxyList.get(block=False)
+                except Exception:
+                    break
+                else:
+                    proxy.conn.close()
+
+        # launcher, run with multiprocessing
+        # def launch(self,taskBuffer,aSiteMapper,holdingAna):
+        def proc_launch(self):
+            # run
+            self.process = multiprocessing.Process(target=self.run)
+            self.process.start()
+
+        # join of multiprocessing
+        def proc_join(self):
+            self.process.join()
+
+
+    # get buildJobs in the holding state
+    tmpLog.debug("get holding build jobs")
+    holdingAna = []
+    varMap = {}
+    varMap[':prodSourceLabel'] = 'panda'
+    varMap[':jobStatus'] = 'holding'
+    status,res = taskBuffer.querySQLS("SELECT PandaID from ATLAS_PANDA.jobsActive4 WHERE prodSourceLabel=:prodSourceLabel AND jobStatus=:jobStatus",varMap)
+    if res is not None:
+        for id, in res:
+            holdingAna.append(id)
+    tmpLog.debug("number of holding Ana %s " % len(holdingAna))
+
+    # add files
+    tmpLog.debug("run Adder processes")
+
+    # make TaskBuffer IF
+    # taskBufferIF = TaskBufferInterface()
+    # taskBufferIF.launch(taskBuffer)
+
+    # p = AdderProcess()
+    # p.run(taskBuffer, aSiteMapper, holdingAna)
+
+    adderThrList = []
+    nThr = 3
```

Comment:
It would be better to use a larger nThr. There is no bad effect even if adder is very quick, and the slowness of adder directly affects the whole system throughput. I see that the node was still acceptable when consuming the backlog with nThr=10

https://monit-grafana.cern.ch/d/000000116/host-metrics?from=now-7d&orgId=1&to=now&var-environment=All&var-hostgroup=vopanda%2Fpandaserver%2Fserver_py3&var-hostname=aipanda056.cern.ch&var-availability_zone=All&var-rp=one_week&var-bin=$__auto_interval_bin

Reply:
About large nThr my concern was not about system usage on nodes, but about too many oracle DB sessions since each process; if nThr=10 on all nodes then ATLAS_PANDA_WRITER will have 300 + sessions ; still below the limit 400 but seems unnecessary and we may need more session in the future.

Anyhow I will try with nThr=6; should be enough I guess.

Reply:
Actually the main reason to use TaskBufferInterface in adder was to reduce the number of connections where the DB connection was established only in the main process which was shared by child processes. I forgot why it cannot be used although a similar mechanism is used in JEDI. Is it because the entire panda-daemon got stuck if TaskBufferInterface was used?

Reply:
It was because at the end of script it called the TaskBufferInterface.terminate():
```
    taskBufferIF.terminate()
```
which had also killed the whole panda-daemon with SIGTERM.

However I just tried add_main with TaskBufferInterface.terminate() again, and it seems OK. Seems I already fixed panda-daemon at some moment, to be immune to signals from child processes and able to keep running:
```
2020-11-16 11:05:03,186 panda.log.daemons: INFO     daemon_loop <worker_pid=2454123> add_main start looping
2020-11-16 11:05:14,978 panda.log.daemons: WARNING  daemon_loop <worker_pid=2454123> (got signal 15)
2020-11-16 11:05:14,995 panda.log.daemons: INFO     daemon_loop <worker_pid=2454123> add_main finish looping
```

I still do not like the SIGTERM from TaskBufferInterface (not sure if it can be terminated gracefully without signals), but if you think it ok I'll add them back.

Reply:
> taskBufferIF.terminate()

This is not required as long as the taskbufferIF process is terminated and doesn't become orphaned or zombie
when the parent exits.